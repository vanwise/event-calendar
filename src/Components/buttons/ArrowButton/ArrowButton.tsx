import styled from 'styled-components/macro';
import { RightArrowIcon } from 'Components/svg';

interface ArrowButtonProps extends ButtonPropsWithoutRef {
  direction?: ArrowDirection;
}

const directionRotateValue = {
  bottom: 90,
  left: 180,
  top: 270,
};

type ArrowDirection = keyof typeof directionRotateValue;

function ArrowButton({ direction, ...props }: ArrowButtonProps) {
  return (
    <Root $direction={direction} type="button" {...props}>
      <Icon />
    </Root>
  );
}

const Root = styled.button<{ $direction?: ArrowDirection }>`
  width: 32px;
  height: 32px;
  padding: 2px;
  transition: 0.3s ease-out;
  background: var(--violet);
  border-radius: 12px;

  ${({ $direction }) =>
    $direction && `transform: rotate(${directionRotateValue[$direction]}deg);`}

  &:hover:not([disabled]) {
    border-radius: 50%;
  }

  &:disabled {
    opacity: 0.5;
  }
`;

const Icon = styled(RightArrowIcon)`
  fill: white;
`;

export default ArrowButton;
