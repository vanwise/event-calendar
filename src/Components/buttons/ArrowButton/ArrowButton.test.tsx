import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ArrowButton, {
  ArrowButtonDirection,
  arrowButtonRotateValue,
} from './ArrowButton';

const onClick = jest.fn();

function setup(props = {}) {
  const renderResult = render(<ArrowButton onClick={onClick} {...props} />);
  const arrowButton = renderResult.getByRole('button');

  return [arrowButton, renderResult] as const;
}

afterEach(jest.clearAllMocks);

describe('Arrow Button', () => {
  it('should be clicked', async () => {
    const [arrowButton] = setup();

    await userEvent.click(arrowButton);
    expect(onClick).toBeCalledTimes(1);
  });

  it('should take right direction', () => {
    const [arrowButton, { rerender }] = setup();

    for (const direction in arrowButtonRotateValue) {
      rerender(<ArrowButton direction={direction as ArrowButtonDirection} />);
      expect(arrowButton).toMatchSnapshot();
    }
  });
});
