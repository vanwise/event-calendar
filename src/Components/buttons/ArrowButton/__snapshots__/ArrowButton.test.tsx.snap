// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow Button should take right direction 1`] = `
.c0 {
  width: 32px;
  height: 32px;
  padding: 2px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  background: var(--violet);
  border-radius: 12px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c0:hover:not([disabled]) {
  border-radius: 50%;
}

.c0:disabled {
  opacity: 0.5;
}

.c1 {
  fill: white;
}

<button
  class="c0"
  type="button"
>
  <svg
    class="c1"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17.92 11.62a1 1 0 0 0-.21-.33l-5-5a1 1 0 0 0-1.42 1.42l3.3 3.29H7a1 1 0 0 0 0 2h7.59l-3.3 3.29a1 1 0 0 0 0 1.42 1 1 0 0 0 1.42 0l5-5a1 1 0 0 0 .21-.33 1 1 0 0 0 0-.76Z"
    />
  </svg>
</button>
`;

exports[`Arrow Button should take right direction 2`] = `
.c0 {
  width: 32px;
  height: 32px;
  padding: 2px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  background: var(--violet);
  border-radius: 12px;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c0:hover:not([disabled]) {
  border-radius: 50%;
}

.c0:disabled {
  opacity: 0.5;
}

.c1 {
  fill: white;
}

<button
  class="c0"
  type="button"
>
  <svg
    class="c1"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17.92 11.62a1 1 0 0 0-.21-.33l-5-5a1 1 0 0 0-1.42 1.42l3.3 3.29H7a1 1 0 0 0 0 2h7.59l-3.3 3.29a1 1 0 0 0 0 1.42 1 1 0 0 0 1.42 0l5-5a1 1 0 0 0 .21-.33 1 1 0 0 0 0-.76Z"
    />
  </svg>
</button>
`;

exports[`Arrow Button should take right direction 3`] = `
.c0 {
  width: 32px;
  height: 32px;
  padding: 2px;
  -webkit-transition: 0.3s ease-out;
  transition: 0.3s ease-out;
  background: var(--violet);
  border-radius: 12px;
  -webkit-transform: rotate(270deg);
  -ms-transform: rotate(270deg);
  transform: rotate(270deg);
}

.c0:hover:not([disabled]) {
  border-radius: 50%;
}

.c0:disabled {
  opacity: 0.5;
}

.c1 {
  fill: white;
}

<button
  class="c0"
  type="button"
>
  <svg
    class="c1"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M17.92 11.62a1 1 0 0 0-.21-.33l-5-5a1 1 0 0 0-1.42 1.42l3.3 3.29H7a1 1 0 0 0 0 2h7.59l-3.3 3.29a1 1 0 0 0 0 1.42 1 1 0 0 0 1.42 0l5-5a1 1 0 0 0 .21-.33 1 1 0 0 0 0-.76Z"
    />
  </svg>
</button>
`;
